### PR TITLE
langchain[patch]: add tool messages formatter for tool calling agent

### PR DESCRIPTION
**Pull request by mackong**
- **Description:** add tool_messages_formatter for tool calling agent, make tool messages can be formatted in different ways for your LLM.
  - **Issue:** N/A
  - **Dependencies:** N/A

### CODE REVIEW

File: libs/langchain/langchain/agents/tool_calling_agent/base.py [hunk 1]
```diff
@@ -21,6 +25,8 @@ def create_tool_calling_agent(
         tools: Tools this agent has access to.
         prompt: The prompt to use. See Prompt section below for more on the expected
             input variables.
+        tool_messags_formatter: Formatter function to convert (AgentAction, tool output)
```
Comment (isahers1) [inline libs/langchain/langchain/agents/tool_calling_agent/base.py:28]: typo

File: libs/langchain/langchain/agents/format_scratchpad/tools.py [hunk 2]
```diff
@@ -10,10 +10,10 @@
 
 from langchain.agents.output_parsers.tools import ToolAgentAction
 
+ToolMessagesFormatter = Callable[[Sequence[Tuple[AgentAction, str]]], List[BaseMessage]]
 
-def _create_tool_message(
-    agent_action: ToolAgentAction, observation: str
-) -> ToolMessage:
+
+def create_tool_message(agent_action: ToolAgentAction, observation: str) -> ToolMessage:
```
Comment (isahers1) [inline libs/langchain/langchain/agents/format_scratchpad/tools.py:16]: why did you mark this as public?

**Comment by isahers1** [comment]
@baskaryan left two small comments, but lgtm other than that

File: libs/langchain/langchain/agents/tool_calling_agent/base.py [hunk 3]
```diff
@@ -6,13 +6,17 @@
...
 )
 from langchain.agents.output_parsers.tools import ToolsAgentOutputParser
 
 
 def create_tool_calling_agent(
-    llm: BaseLanguageModel, tools: Sequence[BaseTool], prompt: ChatPromptTemplate
+    llm: BaseLanguageModel,
+    tools: Sequence[BaseTool],
+    prompt: ChatPromptTemplate,
+    tool_messages_formatter: ToolMessagesFormatter = format_to_tool_messages,
```
Comment (eyurtsev) [inline libs/langchain/langchain/agents/tool_calling_agent/base.py:19]: Please use `*` prior to all named arguments with defaults

```suggestion
    *,
    tool_messages_formatter: ToolMessagesFormatter = format_to_tool_messages,
```

File: libs/langchain/langchain/agents/format_scratchpad/tools.py [hunk 4]
```diff
@@ -10,10 +10,10 @@
 
 from langchain.agents.output_parsers.tools import ToolAgentAction
 
+ToolMessagesFormatter = Callable[[Sequence[Tuple[AgentAction, str]]], List[BaseMessage]]
```
Comment (eyurtsev) [inline libs/langchain/langchain/agents/format_scratchpad/tools.py:13]:
```suggestion
```

File: libs/langchain/langchain/agents/tool_calling_agent/base.py (hunk 3 above)
Comment (eyurtsev) [inline libs/langchain/langchain/agents/tool_calling_agent/base.py:19]: We need a better name for this argument.

Here are some options:

* `message_formatter`
* `history_to_message_formatter`
* `agent_history_to_messages`

File: libs/langchain/langchain/agents/format_scratchpad/tools.py (hunk 4 above)
Comment (eyurtsev) [reply libs/langchain/langchain/agents/format_scratchpad/tools.py:13]: Suggesting to either inline the type in the function OR move it to base.py and update the type name to be better reflect what it does (match the variable name)

File: libs/langchain/langchain/agents/format_scratchpad/tools.py (hunk 2 above)
Comment (mackong) [reply libs/langchain/langchain/agents/format_scratchpad/tools.py]: Before i think it's a good utility function for custom formatter, seems not necessary after think twice.

**Pull request merged**